File: src/main/java/no/nav/nare/core/evaluation/Result.java
package no.nav.nare.core.evaluation;


public enum Result {

    YES(-1, "NO"),
    NO(1, "YES"),
    MAYBE(0, "MAYBE");

    private final int weight;
    private String inverse;


    Result(int weight, String inverse) {
        this.weight = weight;
        this.inverse = inverse;
    }

    public Result and(Result result) {
        return (this.weight > result.weight) ? this : result;
    }

    public Result or(Result result) {
        return (this.weight < result.weight) ? this : result;
    }

    public Result not(){
        return Result.valueOf(this.inverse);

    }
}

File: src/main/kotlin/no/nav/knare/core/evaluations/Result.kt
package no.nav.knare.core.evaluations

enum class Result {
    YES {
        override fun and(other: Result): Result = other
        override fun or(other: Result): Result = YES
        override fun not(): Result = NO
    },

    NO {
        override fun and(other: Result): Result = NO
        override fun or(other: Result): Result = other
        override fun not(): Result = YES
    },

    MAYBE {
        override fun and(other: Result): Result = if (other == YES) MAYBE else other
        override fun or(other: Result): Result = if (other == NO) MAYBE else other
        override fun not(): Result = MAYBE
    };

    abstract fun and(other: Result): Result
    abstract fun or(other: Result): Result
    abstract fun not(): Result
}